site/app/controllers/student/SubmissionController.php

```diff
@@ -1020,6 +1020,63 @@ public function ajaxRegrade($gradeable_id): array {
         return $this->core->getOutput()->renderResultMessage($msg, true);
     }
 
+    /**
+     * @return JsonResponse
+     * @param string $gradeable_id
+     */
+    #[Route("/api/{_semester}/{_course}/gradeable/{gradeable_id}/score", methods: ["GET"])]
+    public function ajaxGetGradeableScore(string $gradeable_id): JsonResponse {
+        // Instructor
+        if ($this->core->getUser()->getAccessLevel() !== 1) {
+            if (($_GET['user_id'] ?? '') !== $this->core->getUser()->getId()) {
+                return JsonResponse::getFailResponse('API key and specified user_id are not for the same user.');
+            }
+        }
+        try {
+            $gradeable = $this->core->getQueries()->getGradeableConfig($gradeable_id);
+        }
+        catch (\InvalidArgumentException $e) {
+            return JsonResponse::getFailResponse('Gradeable does not exist');
+        }
+        $gradeable_version_instance = $this->core->getQueries()->getGradedGradeable(
+            $gradeable,
+            $_GET['user_id'],
+            $gradeable->isTeamAssignment()
+        )->getAutoGradedGradeable()->getActiveVersionInstance();
+
+        if ($gradeable_version_instance === null || $gradeable_version_instance->isQueued() || $gradeable_version_instance->isGrading()) {
+            // Return null value for not graded, or not submitted gradeables.
+            return JsonResponse::getSuccessResponse(null);
+        }
+
+        return JsonResponse::getSuccessResponse($gradeable_version_instance->getTotalPoints());
+    }
+
+    /**
+     * @return JsonResponse|array{
+     *     status: string,
+     *     data: mixed
+     * }
+     */
+    #[Route("/api/{_semester}/{_course}/gradeable/{gradeable_id}/grade", methods: ["POST"])]
+    public function ajaxRequestGrade(string $gradeable_id): JsonResponse|array {
+        // Instructor
+        if ($this->core->getUser()->getAccessLevel() !== 1) {
+            if (($_POST['user_id'] ?? '') !== $this->core->getUser()->getId()) {
+                return JsonResponse::getFailResponse('API key and specified user_id are not for the same user.');
+            }
+        }
+        $vcs_checkout = array_key_exists('vcs_checkout', $_POST) ? $_POST['vcs_checkout'] === "true" : false;
+        if (!$vcs_checkout) {
+            return JsonResponse::getFailResponse("API only supports requesting for VCS gradeables to be graded.");
+        }
+        if (!array_key_exists('git_repo_id', $_POST)) {
+            return JsonResponse::getFailResponse("API requires git_repo_id variable to be set.");
+        }
+
+        return $this->ajaxUploadSubmission($gradeable_id);
```

Comment:
Why doesn't this return a regular json success response?

Reply:
This returns the response from the ajaxUploadSubmission function, which is in the long run a JsonResponse success or fail response.

Reply:
Why does the return type have array an option?

Reply:
This function calls `uploadResult` which then calls `renderResultMessage`, which returns an array formatted as a success or fail JSON response, but as an array.